SignalShareExtension/ShareViewController.swift

```diff
@@ -664,10 +673,75 @@ public class ShareViewController: UIViewController, ShareViewDelegate, SAEFailed
         // FIXME: For now, we prefer the URL provider and discard the text provider, since it's more useful to share the URL than the caption
         // but we *should* include both. This will be a bigger change though since our share extension is currently heavily predicated
         // on one itemProvider per share.
-        guard let itemProvider: NSItemProvider = type(of: self).preferredItemProvider(inputItem: inputItem) else {
-            let error = ShareViewControllerError.assertionError(description: "No item provider in input item attachments")
+
+        // Prefer a URL provider if available
+        if let preferredAttachment = attachments.first(where: { (attachment: Any) -> Bool in
+            guard let itemProvider = attachment as? NSItemProvider else {
+                return false
+            }
+            return isUrlItem(itemProvider: itemProvider)
+        }) {
+            if let itemProvider = preferredAttachment as? NSItemProvider {
+                return [itemProvider]
+            } else {
+                owsFailDebug("Unexpected attachment type: \(String(describing: preferredAttachment))")
+            }
+        }
+
+        // else return whatever is available
+        if let itemProvider = inputItem.attachments?.first as? NSItemProvider {
+            return [itemProvider]
+        } else {
+            owsFailDebug("Missing attachment.")
+        }
+        return []
+    }
+
+    private func selectItemProviders() -> Promise<[NSItemProvider]> {
```

Comment:
I did it this way to streamline the error handling.  Otherwise, we'd have to modify `buildAttachments()` to catch the exception turn it into a error promise, etc.